libbeat/esclientleg/connection.go

```diff
@@ -0,0 +1,235 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package esclientleg
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+
+	"github.com/elastic/beats/libbeat/logp"
+
+	"github.com/elastic/beats/libbeat/common"
+)
+
+// Connection manages the connection for a given client.
+type Connection struct {
+	ConnectionSettings
+
+	version common.Version
+	log     *logp.Logger
+}
+
+// ConnectionSettings are the settings needed for a Connection
+type ConnectionSettings struct {
+	URL      string
+	Username string
+	Password string
+	APIKey   string
+	Headers  map[string]string
+
+	HTTP              *http.Client
+	OnConnectCallback func() error
+
+	Encoder BodyEncoder
+}
+
+func NewConnection(settings ConnectionSettings) *Connection {
+	return &Connection{
+		ConnectionSettings: settings,
+		log:                logp.NewLogger("esclientleg"),
+	}
+}
+
+// Connect connects the client. It runs a GET request against the root URL of
+// the configured host, updates the known Elasticsearch version and calls
+// globally configured handlers.
+func (conn *Connection) Connect() error {
```

Comment:
For now we need to keep `Connect` here I guess. But for the future API based on top of go-elasticsearch, I think it would be nice to not have a Connect method that allows us to reconnect. Instead provide a `Dial` (or `Connect`) function that returns a `Conn` object. Once closed we would not have 'reconnect' support anymore then, but would require to dial/connect again, constructing a very new object. Based on this we might maintain a connection pool.